### PR TITLE
✨ enables explicit disabling of cluster module

### PR DESCRIPTION
**Pull request by RenilRaj-BR**
What this PR does:

The cluster module is currently enabled for all vSphere clusters based on the anti-affinity feature flag, with no option to disable its creation if the user doesn't want this feature. This PR introduces a DisableClusterModule field to the vSphereCluster API, allowing users to control the creation of a cluster module. However, this field is only applicable when the NodeAntiAffinity feature flag is enabled; otherwise, it is ignored.

### CODE REVIEW

File: apis/v1beta1/vspherecluster_types.go [hunk 1]
```diff
@@ -59,6 +59,12 @@ type VSphereClusterSpec struct {
 	// +optional
 	ClusterModules []ClusterModule `json:"clusterModules,omitempty"`
 
+	// DisableClusterModule is used to explicitly turn off the ClusterModule feature.
+	// This should work along side NodeAntiAffinity feature flag.
+	// If the NodeAntiAffinity feature flag is turned off, this will be disregarded.
```
Comment (neolit123) [inline apis/v1beta1/vspherecluster_types.go:64]: this is fine but has to be updated once the feature flag graduates and is removed

**Review by neolit123 — COMMENTED** [review]
/ok-to-test

**Comment by rikatz** [comment]
@RenilRaj-BR test failures are legit.

If you look into conversion there is just the entry of // WARNING: in.DisableClusterModule requires manual conversion: does not exist in peer-type 

You need to establish some manual conversion for this field to not add it to the v1alpha3 and v1alpha4 API.

File: apis/v1beta1/vspherecluster_types.go (hunk 1 above)
Comment (RenilRaj-BR) [reply apis/v1beta1/vspherecluster_types.go:64]: yeah. Will do it once it is graduated

**Review by rgcostea — APPROVED** [review]

**Comment by rikatz** [comment]
/lgtm
/approve

Not that my approval counts much :) But I think this is a good one to go!

@sbueringer @fabriziopandini if you have some time, this is a "small" one :)

**Comment by rikatz** [comment]
https://github.com/team-cluster-api/cluster-api-provider-vsphere/pull/36 for e2e tests

File: controllers/vspherecluster_reconciler.go [hunk 2]
```diff
@@ -359,7 +359,7 @@ func (r *clusterReconciler) reconcileDeploymentZones(ctx context.Context, cluste
 }
 
 func (r *clusterReconciler) reconcileClusterModules(ctx context.Context, clusterCtx *capvcontext.ClusterContext) (reconcile.Result, error) {
-	if feature.Gates.Enabled(feature.NodeAntiAffinity) {
+	if feature.Gates.Enabled(feature.NodeAntiAffinity) && !clusterCtx.VSphereCluster.Spec.DisableClusterModule {
```
Comment (sbueringer) [inline controllers/vspherecluster_reconciler.go:362]: What about the other places where the feature gate is used?

Comment (RenilRaj-BR) [reply controllers/vspherecluster_reconciler.go:362]: @sbueringer,
Added the missing condition in the vspherevm_controller.go.
Didn't add the missing condition in `AddClusterControllerToManager` function in the vspherecluster_controller.go as this is a one time config and can not be changed per cluster

Comment (sbueringer) [reply controllers/vspherecluster_reconciler.go:362]: Makes sense, thx!

**Comment by sbueringer** [comment]
Thx folks!!

/lgtm
/approve

**Pull request merged**